src/components/map-controls.js

```diff
@@ -122,79 +143,98 @@ export default class MapControls extends PureComponent {
     ));
   }
 
-  _onTouchRotate(opts) {
-    this._onMouseRotate(opts);
-  }
-
-  _onMouseDown({pos}) {
+  _onDragStart(event) {
+    const pos = this._getCenter(event);
     const newMapState = this.props.mapState.panStart({pos}).rotateStart({pos});
     this._updateViewport(newMapState, {isDragging: true});
   }
 
-  _onMouseDrag({pos, startPos, modifier}) {
-    if (!this.props.onChangeViewport) {
-      return;
-    }
-
-    if (modifier) {
-      this._onMouseRotate({pos, startPos});
+  _onDrag(event) {
+    if (this._isFunctionKeyPressed(event)) {
+      this._onRotate(event);
     } else {
-      this._onMousePan({pos});
+      this._onPan(event);
     }
   }
 
-  _onMousePan({pos}) {
+  _onDragEnd(event) {
+    const newMapState = this.props.mapState.panEnd().rotateEnd();
+    this._updateViewport(newMapState, {isDragging: false});
+  }
+
+  _onPan(event) {
+    const pos = this._getCenter(event);
     const newMapState = this.props.mapState.pan({pos});
     this._updateViewport(newMapState);
   }
 
-  _onMouseRotate({pos, startPos}) {
-    if (!this.props.onChangeViewport || !this.props.perspectiveEnabled) {
+  _onRotate(event) {
+    if (!this.props.perspectiveEnabled) {
       return;
     }
 
-    const xDelta = pos[0] - startPos[0];
-    const yDelta = pos[1] - startPos[1];
+    const {deltaX, deltaY} = event;
+    const pos = this._getCenter(event);
+    const startPos = [pos[0] - deltaX, pos[1] - deltaY];
 
-    const xDeltaScale = xDelta / this.props.width;
+    const xDeltaScale = deltaX / this.props.width;
```

Comment:
nit: since we're switching to having X/Y at the end, let's get these guys to: `deltaScaleX, deltaScaleY`